### PR TITLE
`get_queued` should respect {start,end}_delivery_time

### PR DESCRIPTION
**Pull request by SeiryuZ**
3rd Step for #31

### CODE REVIEW

File: sms_engine/sms.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 
 def create(to=None, message="", description="", scheduled_time=None, priority=None,
-           commit=True, backend=""):
+           commit=True, backend="", start_delivery_time=None, end_delivery_time=None):
```
Comment (selwin) [inline sms_engine/sms.py]: `delivery_window=(start, end)`

File: sms_engine/sms.py [hunk 2]
```diff
@@ -60,9 +64,23 @@ def send(to=None, message="", description="", scheduled_time=None, priority=None
 
 def get_queued():
     limit = settings.SMS_ENGINE.get('BATCH_SIZE', 50)
-    return SMS.objects.filter(status=STATUS.queued) \
-        .filter(Q(scheduled_time__lte=timezone.now()) | Q(scheduled_time=None))\
-        .order_by('-priority')[:limit]
+
+    try:
+        now = timezone.localtime()
+    except ValueError:
```
Comment (selwin) [inline sms_engine/sms.py]: Kenapa bisa `ValueError`?

Comment (SeiryuZ) [reply sms_engine/sms.py]: kalau `USE_TZ` nya `False` di settings, dia throw `ValueError` di python3, di python2 dia throw `TypeError`

File: sms_engine/tests/test_sms.py [hunk 3]
```diff
@@ -1,13 +1,65 @@
...
+from django.test import TestCase, override_settings
+from django.utils import timezone
+
 from sms_engine.models import SMS, PRIORITY, STATUS, Log
 from sms_engine.sms import create as create_sms, send as send_sms, _send_bulk, get_queued
 
 
 class SMSTest(TestCase):
 
+    def test_get_queued_sms_respecting_delivery_window(self):
```
Comment (selwin) [inline sms_engine/tests/test_sms.py]: `test_delivery_window`

File: sms_engine/tests/test_sms.py [hunk 4]
```diff
@@ -1,13 +1,65 @@
...
+
+        # 7 AM
+        with freeze_time("2000-1-1 00:00", tz_offset=7):
+            sms1 = create_sms(to="+62800000000001", message="Test", backend='dummy')
+            sms2 = create_sms(to="+62800000000001", message="Test", backend='dummy',
+                              start_delivery_time=time(10, 0), end_delivery_time=time(22, 0))
+            self.assertEqual([sms1], list(get_queued()))
```
Comment (selwin) [inline sms_engine/tests/test_sms.py]: Coba `freeze_time` ke jam 7 dan jam 10 untuk cek apakah sms2 dapet.

File: sms_engine/sms.py [hunk 5]
```diff
@@ -16,20 +16,30 @@
...
+        delivery_window - Tuple representing start and end of delivery time
     """
     priority = parse_priority(priority)
     status = None if priority == PRIORITY.now else STATUS.queued
 
     if backend not in get_available_backends().keys():
         raise ValueError('%s is not a valid backend alias' % backend)
 
+    if delivery_window:
+        start = delivery_window[0]
```
Comment (selwin) [inline sms_engine/sms.py]: Perlu verify kalau start < end

File: sms_engine/sms.py [hunk 6]
```diff
@@ -60,9 +70,23 @@ def send(to=None, message="", description="", scheduled_time=None, priority=None
 
 def get_queued():
     limit = settings.SMS_ENGINE.get('BATCH_SIZE', 50)
-    return SMS.objects.filter(status=STATUS.queued) \
-        .filter(Q(scheduled_time__lte=timezone.now()) | Q(scheduled_time=None))\
-        .order_by('-priority')[:limit]
+
+    try:
+        now = timezone.localtime()
+    except (ValueError, TypeError):
```
Comment (selwin) [inline sms_engine/sms.py]: I think di sini bisa pake `timezone.now()` aja, since udah ngebalikkin timezone.

Comment (SeiryuZ) [reply sms_engine/sms.py]: Lo yakin? kalau timezone.now dia balikin timezone tapi dalam UTC, yang kemungkinan bukan yg user mau, dan query window time ini jadi lebih complicated

Screen Shot 2019-05-09 at 10 01 51 AM

File: sms_engine/sms.py [hunk 7]
```diff
@@ -16,20 +17,37 @@
...
 
     if backend not in get_available_backends().keys():
         raise ValueError('%s is not a valid backend alias' % backend)
 
+    if delivery_window:
+        start = delivery_window[0]
+        end = delivery_window[1]
+
+        # Validate start + end
+        if type(start) != time or type(end) != time:
```
Comment (selwin) [inline sms_engine/sms.py]: Pake `isinstance`

File: sms_engine/tests/test_sms.py [hunk 8]
```diff
@@ -1,13 +1,74 @@
...
+        with freeze_time("2000-1-1 00:00:00", tz_offset=7):
+            sms1 = create_sms(to="+62800000000001", message="Test", backend='dummy')
+            sms2 = create_sms(to="+62800000000001", message="Test", backend='dummy',
+                              delivery_window=(time(10, 0), time(22, 0)))
+            self.assertEqual([sms1], list(get_queued()))
+
+        # 10 AM
+        SMS.objects.all().delete()
+        with freeze_time("2000-1-1 03:00:00", tz_offset=7):
+            sms1 = create_sms(to="+62800000000001", message="Test", backend='dummy')
```
Comment (selwin) [inline sms_engine/tests/test_sms.py]: Nggak usah create SMS object baru.

Comment (selwin) [reply sms_engine/tests/test_sms.py]: Tambahin check jam 11 malem juga.

File: sms_engine/sms.py [hunk 9]
```diff
@@ -16,20 +17,37 @@
...
+        delivery_window - Tuple representing start and end of delivery time
     """
     priority = parse_priority(priority)
     status = None if priority == PRIORITY.now else STATUS.queued
 
     if backend not in get_available_backends().keys():
         raise ValueError('%s is not a valid backend alias' % backend)
 
+    if delivery_window:
+        start = delivery_window[0]
```
Comment (selwin) [inline sms_engine/sms.py]: Kalo pake `delivery_window`, `USE_TZ` must be True

File: sms_engine/sms.py [hunk 10]
```diff
@@ -34,9 +34,12 @@ def create(to=None, message="", description="", scheduled_time=None, priority=No
         end = delivery_window[1]
 
         # Validate start + end
-        if type(start) != time or type(end) != time:
+        if not isinstance(start, time) or not isinstance(end, time):
             raise ValueError('start/end is not a valid time type')
 
+        if not settings.USE_TZ:
+            raise ValueError('Delivery Window is only supported for Django settings with USE_TZ=True')
```
Comment (selwin) [inline sms_engine/sms.py]: `delivery_window is only supported for projects with ``USE_TZ = True`` `

**Pull request merged**